src/main/java/com/google/starfish/servlets/SearchServlet.java

```diff
@@ -0,0 +1,94 @@
+import java.util.ArrayList;
+
+/** Servlet that returns search results for notes based on school and course */
+@WebServlet("/search")  
+public class SearchServlet extends HttpServlet {  
+
+  private NoteService noteService = new NoteService();
+
+  @Override
+  public void doGet(HttpServletRequest req, HttpServletResponse res) throws ServletException, IOException {
+    DataSource pool = (DataSource) req.getServletContext().getAttribute("my-pool");  
+    String reqSchool = req.getParameter("school").toLowerCase().trim();
+    String reqCourse= req.getParameter("course").toLowerCase().trim();
+    ArrayList<Note> notes = new ArrayList<>();
+    try (Connection conn = pool.getConnection()) {
+      String stmt = getSearchQuery(reqSchool, reqCourse);
+      try (PreparedStatement getNotesStatement = conn.prepareStatement(stmt)) {
+        getNotesStatement.setString(1, reqSchool);
+        getNotesStatement.setString(2, reqCourse);
+        ResultSet rs = getNotesStatement.executeQuery();
+
+        while (rs.next()) {
+          long noteId = rs.getLong("id");
+          long authorId = rs.getLong("author_id");
+          String school = rs.getString("school");
+          String course = rs.getString("course");
+          String title = rs.getString("title");
+          String sourceUrl = rs.getString("title");
+          String pdfSource = rs.getString("pdf_source");
+          Date dateCreated = rs.getDate("date_created");
+          long numDownloads = rs.getLong("num_downloads");
+          long numFavorites = noteService.getNumFavoritesById(pool, noteId);
+
+          Note thisNote = new Note.Builder()
+                              .setId(noteId)
+                              .setAuthorId(authorId)
+                              .setRequiredLabels(school, course)
+                              .setNoteTitle(title)
+                              .setSourceUrl(sourceUrl)
+                              .setOptionalPdfSource(pdfSource)
+                              .setDateCreated(dateCreated)
+                              .setNumDownloads(numDownloads)
+                              .setNumFavorites(numFavorites)
+                              .build();
+          notes.add(thisNote);
+        }
+        String json = convertArrayListToJson(notes);
+        res.setContentType("application/json");
+        res.getWriter().println(json);
+      }
+    } catch (SQLException ex) {
+        res.getWriter().println("Error caught while speaking to database: " + ex);  
+    }
+  }  
+
+  /** Dynamically generates sql query statement for note search */
+  private String getSearchQuery(String school, String course) {
+    String stmt = 
+        "SELECT * "
+      + "FROM `starfish.notes` "
+      + "WHERE 1=1";
+    if (school != null && !school.isEmpty()) {
+      stmt += " AND `school`= ?";
+    }
+    if (course != null && !course.isEmpty()) {
+      stmt += " AND `course`= ?";
+    }
+    stmt += ";";
+    return stmt;
+  }
```

Comment:
I'm sure it'll be easy to add a SORT BY dateCreated or numFavorites piece to the statement, but I'm wondering: if SORT BY dateCreated is not specified, do you think the database would return the notes in Chronological order anyway? Is there an order to how rows are stored? Or would omitting SORT BY leave the order unpredictable?

Reply:
Yeah so we should be able to easily sort by dateCreated or numFavorites but if we don't specify this it's technically undefined, although most likely what will happen is that it returns by time inserted into the database. Like the id field is autoincrementing so technically that's the order in which rows are stored.